<!doctype html>
<html lang="en">
<head>
  <title>BalticMicrobeDB</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <link href="https://fonts.googleapis.com/css?family=Roboto:400,700,400italic|Source+Code+Pro:400,700" rel="stylesheet" type="text/css">
  <link rel="stylesheet"href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap.min.css">
  <link rel="stylesheet" href="{{ url_for('static', filename='main.css') }}?v=0.2">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.2/jquery.min.js"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js"></script>
  <script src="https://code.highcharts.com/highcharts.js"></script>
  <script src="{{ url_for('static', filename='main.js') }}?v=0.2"></script>
</head>
<body>
  <nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <a href='/' class="navbar-brand">BalticMicrobeDB</a>
      </div>
      <ul class="nav navbar-nav navbar-right">
        <li>
          <button type="button"  class="btn btn-info" data-toggle="modal" data-target="#IndexModal">
          <span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> About BalticMicrobeDB
          </button>
        </li>
      </ul>
    </div>
  </nav>

  <!-- Modal -->
  <div class="modal fade" id="IndexModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel">
    <div class="modal-dialog" role="document">
      <div class="modal-content">
        <div class="modal-header">
          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
          <h2 class="modal-title" id="myModalLabel">About BalticMicrobeDB</h2>
        </div>
        <div class="modal-body">
          <p>
            BalticMicrobeDB is a database containing quantifications of genes from the Baltic Sea Reference Metagenome (BARM).
            The quantifications are summarized for both taxonomic and functional annotations and can be compared between samples.
            Furthermore, a BLAST based query can be performed to compare any given sequence to what genes and annotations that are present within BARM.
            Worth noting is that only genes with either taxonomic or a functional anotation is included in the BalticMicrobeDB.
          </p>
          <p>
            Three sample sets are included, a time series, a spatial transect and a redoxcline sample set. The time series originates from the Linnaeus Microbial Observatory, year 2012, and is simply named lmo in the database. The spatial transect, named transect in the database, consists of three samples at different depths for each of ten stations with locations ranging across the entire Baltic Sea. The redoxcline sample set consists of samples from two stations targeting the redoxcline.
          </p>
          <p>
            The genes in the database originates from a combined assembly of these three datasets with the combined size of 2.6 billion reads. The assembly was conducted with Megahit followed by gene prediction with Prodigal and resulted in 6.8 million unique gene sequences. Out of these, 5.4 million genes received some kind of annotation and was included in the BalticMicrobeDB.
          </p>
          <p>
          All data related to BARM, such as gene sequences, taxonomic annotations, functional annotations, and quantifications of individual genes can be downloaded from <a href="https://figshare.com/s/3171d2e7c0f470fc488a">Figshare</a>. Read sequences together with sample metadata can be found on <a href="https://www.ebi.ac.uk/ena/data/view/PRJEB22997">European Nucleotide Archive</a> and the lmo dataset at <a href="https://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=273799">NCBI Short Read Archive</a>.
          </p>
          <p>
            For more detailed information, please refer to the accompanying manuscript (submitted) "BARM and BalticMicrobeDB: a reference metagenome and interface to meta-omic data for the Baltic Sea".
          </p>
        </div>
        <div class="modal-footer">
          <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
        </div>
      </div>
    </div>
  </div>

  <div class="main-content">
    {% with messages = get_flashed_messages(with_categories=true) %}
      {% if messages %}
        <ul class="flash">
        {% for category, message in messages %}
            {% if category == 'error' %}
              <li><p class="alert alert-warning {{ category }}">{{ message }}</p></li>
            {% else %}
              <li><p class="{{ category }}">{{ message }}</p></li>
            {% endif %}
        {% endfor %}
        </ul>
      {% endif %}
    {% endwith %}
    {% block body %}
    {% endblock body %}
  </div>

  <div class="container">
  <hr>

  <footer>
    <p>
      <div class="row">
        <div class="col-md-5">
          <p>&copy; 2016-2017 Johannes Alneberg, Anders F. Andersson and the <a href="http://blueprint-project.org/">BONUS BLUEPRINT Project</a></p>
          <p>The BONUS BLUEPRINT project has received funding from BONUS (Art 185), funded jointly by the EU and the national funding institutions of Denmark, Sweden, Germany, Finland, and Estonia.</p>
        </div>
        <div class="col-md-6 col-md-offset-1">
          <img src="{{ url_for('static', filename='images/eu.png') }}" class="additional-logo">
          <img src="{{ url_for('static', filename='images/bonus.png') }}" class="additional-logo">
          <img src="{{ url_for('static', filename='images/kth.png') }}" class="additional-logo">
          <img src="{{ url_for('static', filename='images/scilifelab_2.png') }}" class="additional-logo vertical-centered">
        </div>
      </div>
    </p>
  </footer>
  </div>
</body>
</html>
